File: backend/app/api/routes/nl2sql.py
"""API routes for Natural Language to SQL feature."""
from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request

from backend.app.services.security import require_api_key
from backend.app.schemas.nl2sql import (
    NL2SQLGenerateRequest,
    NL2SQLExecuteRequest,
    NL2SQLSaveRequest,
)
from backend.app.services.nl2sql.service import NL2SQLService

router = APIRouter(dependencies=[Depends(require_api_key)])


def get_service() -> NL2SQLService:
    return NL2SQLService()


@router.post("/generate")
async def generate_sql(
    payload: NL2SQLGenerateRequest,
    request: Request,
    svc: NL2SQLService = Depends(get_service),
):
    """Generate SQL from a natural language question."""
    correlation_id = getattr(request.state, "correlation_id", None)
    result = svc.generate_sql(payload, correlation_id)
    return {
        "status": "ok",
        "sql": result.sql,
        "explanation": result.explanation,
        "confidence": result.confidence,
        "warnings": result.warnings,
        "original_question": result.original_question,
        "correlation_id": correlation_id,
    }


@router.post("/execute")
async def execute_query(
    payload: NL2SQLExecuteRequest,
    request: Request,
    svc: NL2SQLService = Depends(get_service),
):
    """Execute a SQL query and return results."""
    correlation_id = getattr(request.state, "correlation_id", None)
    result = svc.execute_query(payload, correlation_id)
    return {
        "status": "ok",
        "columns": result.columns,
        "rows": result.rows,
        "row_count": result.row_count,
        "total_count": result.total_count,
        "execution_time_ms": result.execution_time_ms,
        "truncated": result.truncated,
        "correlation_id": correlation_id,
    }


@router.post("/explain")
async def explain_query(
    request: Request,
    sql: str = Query(..., min_length=1, max_length=10000),
    svc: NL2SQLService = Depends(get_service),
):
    """Get a natural language explanation of a SQL query."""
    correlation_id = getattr(request.state, "correlation_id", None)
    explanation = svc.explain_query(sql, correlation_id)
    return {
        "status": "ok",
        "explanation": explanation,
        "correlation_id": correlation_id,
    }


@router.post("/save")
async def save_query(
    payload: NL2SQLSaveRequest,
    request: Request,
    svc: NL2SQLService = Depends(get_service),
):
    """Save a query as a reusable data source."""
    correlation_id = getattr(request.state, "correlation_id", None)
    saved = svc.save_query(payload, correlation_id)
    return {
        "status": "ok",
        "query": saved.model_dump(mode="json"),
        "correlation_id": correlation_id,
    }


@router.get("/saved")
async def list_saved_queries(
    request: Request,
    connection_id: Optional[str] = Query(None, max_length=64),
    tags: Optional[List[str]] = Query(None),
    svc: NL2SQLService = Depends(get_service),
):
    """List saved queries."""
    correlation_id = getattr(request.state, "correlation_id", None)
    queries = svc.list_saved_queries(connection_id=connection_id, tags=tags)
    return {
        "status": "ok",
        "queries": [q.model_dump(mode="json") for q in queries],
        "correlation_id": correlation_id,
    }


@router.get("/saved/{query_id}")
async def get_saved_query(
    query_id: str,
    request: Request,
    svc: NL2SQLService = Depends(get_service),
):
    """Get a saved query by ID."""
    correlation_id = getattr(request.state, "correlation_id", None)
    query = svc.get_saved_query(query_id)
    if not query:
        raise HTTPException(
            status_code=404,
            detail={
                "status": "error",
                "code": "not_found",
                "message": "Query not found",
            },
        )
    return {
        "status": "ok",
        "query": query.model_dump(mode="json"),
        "correlation_id": correlation_id,
    }


@router.delete("/saved/{query_id}")
async def delete_saved_query(
    query_id: str,
    request: Request,
    svc: NL2SQLService = Depends(get_service),
):
    """Delete a saved query."""
    correlation_id = getattr(request.state, "correlation_id", None)
    deleted = svc.delete_saved_query(query_id)
    return {
        "status": "ok" if deleted else "error",
        "deleted": deleted,
        "query_id": query_id,
        "correlation_id": correlation_id,
    }


@router.get("/history")
async def get_query_history(
    request: Request,
    connection_id: Optional[str] = Query(None, max_length=64),
    limit: int = Query(50, ge=1, le=200),
    svc: NL2SQLService = Depends(get_service),
):
    """Get query history."""
    correlation_id = getattr(request.state, "correlation_id", None)
    history = svc.get_query_history(connection_id=connection_id, limit=limit)
    return {
        "status": "ok",
        "history": [h.model_dump(mode="json") for h in history],
        "correlation_id": correlation_id,
    }


@router.delete("/history/{entry_id}")
async def delete_query_history_entry(
    entry_id: str,
    request: Request,
    svc: NL2SQLService = Depends(get_service),
):
    """Delete a query history entry."""
    correlation_id = getattr(request.state, "correlation_id", None)
    deleted = svc.delete_query_history_entry(entry_id)
    return {
        "status": "ok" if deleted else "error",
        "deleted": deleted,
        "entry_id": entry_id,
        "correlation_id": correlation_id,
    }

File: frontend/src/api/nl2sql.js
/**
 * API client for Natural Language to SQL feature
 */
import { api, API_BASE } from './client'

const asArray = (payload, keys = []) => {
  if (Array.isArray(payload)) return payload
  if (!payload || typeof payload !== 'object') return []
  for (const key of keys) {
    if (Array.isArray(payload[key])) return payload[key]
  }
  return []
}

/**
 * Generate SQL from a natural language question
 * @param {Object} params
 * @param {string} params.question - Natural language question
 * @param {string} params.connectionId - Connection ID
 * @param {string[]} [params.tables] - Optional list of tables to query
 * @param {string} [params.context] - Optional additional context
 * @returns {Promise<Object>} Generated SQL with explanation
 */
export async function generateSQL({ question, connectionId, tables, context }) {
  const { data } = await api.post('/nl2sql/generate', {
    question,
    connection_id: connectionId,
    tables: tables?.length ? tables : undefined,
    context: context || undefined,
  })
  return data
}

/**
 * Execute a SQL query
 * @param {Object} params
 * @param {string} params.sql - SQL query to execute
 * @param {string} params.connectionId - Connection ID
 * @param {number} [params.limit=100] - Max rows to return
 * @param {number} [params.offset=0] - Row offset
 * @returns {Promise<Object>} Query results
 */
export async function executeQuery({ sql, connectionId, limit = 100, offset = 0, includeTotal = false }) {
  const { data } = await api.post('/nl2sql/execute', {
    sql,
    connection_id: connectionId,
    limit,
    offset,
    include_total: includeTotal,
  })
  return data
}

/**
 * Get a natural language explanation of a SQL query
 * @param {string} sql - SQL query to explain
 * @returns {Promise<Object>} Explanation
 */
export async function explainQuery(sqlOrConnectionId, maybeSql) {
  const sql = maybeSql ?? sqlOrConnectionId
  const { data } = await api.post('/nl2sql/explain', null, {
    params: { sql },
  })
  return data
}

/**
 * Save a query as a reusable data source
 * @param {Object} params
 * @param {string} params.name - Name for the saved query
 * @param {string} params.sql - SQL query
 * @param {string} params.connectionId - Connection ID
 * @param {string} [params.description] - Optional description
 * @param {string} [params.originalQuestion] - Original NL question
 * @param {string[]} [params.tags] - Optional tags
 * @returns {Promise<Object>} Saved query
 */
export async function saveQuery({ name, sql, connectionId, description, originalQuestion, tags }) {
  const { data } = await api.post('/nl2sql/save', {
    name,
    sql,
    connection_id: connectionId,
    description: description || undefined,
    original_question: originalQuestion || undefined,
    tags: tags?.length ? tags : undefined,
  })
  return data
}

/**
 * List saved queries
 * @param {Object} [params]
 * @param {string} [params.connectionId] - Filter by connection
 * @param {string[]} [params.tags] - Filter by tags
 * @returns {Promise<Object>} List of saved queries
 */
export async function listSavedQueries({ connectionId, tags } = {}) {
  const { data } = await api.get('/nl2sql/saved', {
    params: {
      connection_id: connectionId || undefined,
      tags: tags?.length ? tags : undefined,
    },
  })
  if (Array.isArray(data)) {
    return { queries: data, total: data.length }
  }
  if (data && typeof data === 'object') {
    const queries = asArray(data, ['queries', 'saved_queries', 'items', 'results'])
    return { ...data, queries, total: data.total ?? data.count ?? queries.length }
  }
  return { queries: [], total: 0 }
}

/**
 * Get a saved query by ID
 * @param {string} queryId - Query ID
 * @returns {Promise<Object>} Saved query
 */
export async function getSavedQuery(queryId) {
  const { data } = await api.get(`/nl2sql/saved/${encodeURIComponent(queryId)}`)
  if (data && typeof data === 'object' && data.query) {
    return data.query
  }
  return data
}

/**
 * Delete a saved query
 * @param {string} queryId - Query ID
 * @returns {Promise<Object>} Deletion result
 */
export async function deleteSavedQuery(queryId) {
  const { data } = await api.delete(`/nl2sql/saved/${encodeURIComponent(queryId)}`)
  return data
}

/**
 * Get query history
 * @param {Object} [params]
 * @param {string} [params.connectionId] - Filter by connection
 * @param {number} [params.limit=50] - Max entries to return
 * @returns {Promise<Object>} Query history
 */
export async function getQueryHistory(params = {}) {
  const options = typeof params === 'string'
    ? { connectionId: params, limit: 50 }
    : (params || {})
  const { connectionId, limit = 50 } = options
  const { data } = await api.get('/nl2sql/history', {
    params: {
      connection_id: connectionId || undefined,
      limit,
    },
  })
  if (Array.isArray(data)) {
    return { history: data, total: data.length }
  }
  if (data && typeof data === 'object') {
    const history = asArray(data, ['history', 'queries', 'items', 'results'])
    return { ...data, history, total: data.total ?? data.count ?? history.length }
  }
  return { history: [], total: 0 }
}

/**
 * Delete a query history entry
 * @param {string} entryId - History entry ID
 * @returns {Promise<Object>} Deletion result
 */
export async function deleteQueryHistoryEntry(entryId) {
  const { data } = await api.delete(`/nl2sql/history/${encodeURIComponent(entryId)}`)
  return data
}
